shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -205,10 +203,38 @@ final class TsTypegenCodeBuilder {
     classDeclaration ++= SourceCode.space ++ SourceCode.openCurlyBrace
 
     // add body fields
-    bodyFieldAndTypes.iterator.foreach({ case (fieldVar, fieldType) => {
-      classDeclaration += SourceCode("    ") ++ fieldVar ++ SourceCode.colon ++ fieldType }})
+    classFieldsAndType
+      .map {case (fieldVar, fieldTypes) => 
+        // field types will always have 1 or more elements
+        val fieldTypesCode = if (fieldTypes.length === 1) {
+          toTsType(fieldTypes(0))(TypegenContext(fieldTypes(0)), Some(true))
+        } else {
+          // multiple types are unioned hence using union precedence
+          // however only consider unique types for union
+          val fieldTypesCode = fieldTypes.toSet[Type].map(fieldType => toTsType(fieldType, false, 1)(TypegenContext(fieldType), Some(true))).toList
+          SourceCode.sepBy(fieldTypesCode, SourceCode.ampersand)
+        }
+        (SourceCode(fieldVar.name), fieldTypesCode)
+      }
+      .foreach({ case (fieldVar, fieldType) => {
+        classDeclaration += SourceCode("    ") ++ fieldVar ++ SourceCode.colon ++ fieldType
+      }})
+
     // constructor needs all fields including super classes
-    val allFieldsAndTypes = bodyFieldAndTypes ++ baseClass.map(getClassFieldAndTypes(_, true)).getOrElse(List.empty)
+    val allFieldsAndTypes = (classFieldsAndType ++
+      baseClass.map(getClassFieldAndTypes(_, true)).getOrElse(List.empty).groupMap(_._1)(_._2))
+      .map {case (fieldVar, fieldTypes) => 
+        // field types will always have 1 or more elements
+        val fieldTypesCode = if (fieldTypes.length === 1) {
+          toTsType(fieldTypes(0))(TypegenContext(fieldTypes(0)), Some(true))
+        } else {
+          // multiple types are unioned hence using union precedence
+          // however only consider unique types for union
+          val fieldTypesCode = fieldTypes.toSet[Type].map(fieldType => toTsType(fieldType, false, 1)(TypegenContext(fieldType), Some(true))).toList
+          SourceCode.sepBy(fieldTypesCode, SourceCode.ampersand)
```

Comment:
Nitpick: `&` is intersection, not union.

Reply:
:grimacing: my bad mixed this up. Fixed.